pkg/menus/menus.go

```diff
@@ -0,0 +1,146 @@
+type Entry interface {
+    //label is the string will show in menu
+	Label() string 
+    //IsDefault is true means this entry will be hit by default, if there is many default choise, the first on in the list will be choose. the rest will be ignored. 
+    IsDefault() bool  
+}
+
+
+//GetChoose will present all entries into a menu with numbers, so that user can choose from them.
+func GetChoose(coord *int, introwords string, entries []Entry) (int, error){
+	//open a new window
+	if err := ui.Init(); err != nil {
+        log.Fatalf("failed to initialize termui: %v", err)
+    }
+    listData := []string{}//listData contains all choice's labels
+    
+    for i, e := range entries {
+        listData = append(listData, fmt.Sprintf("[%d] %s", i+1, e.Label()))
+	}
+
+    //l is a List item which will present like a menu by termui package
+    l := widgets.NewList()	
+    l.Title = "Menu"
+	l.Rows = listData
+    // design the size and location of the menu
+	l.SetRect(0, *coord, 50, *coord+2+len(entries)) 
+    *coord += 2+len(entries)
+	l.TextStyle.Fg = ui.ColorWhite
+    ui.Render(l)      
+    log.Printf("\n"+introwords+"\n\n")
+
+    //p1 is a paragraph for user to input their choice, but it is also used for showing warnings
+    p1 := widgets.NewParagraph() 
+	p1.Text = ""
+	p1.Border = false
+	p1.SetRect(0, *coord, 50, *coord+4) 
+    *coord += 4
+    p1.TextStyle.Fg = ui.ColorWhite
+    ui.Render(p1)      
+    
+    tempFlag := false
+    // keep tracking all input from user
+	uiEvents := ui.PollEvents() 
+    for {
+        e := <-uiEvents
+        if(e.Type != ui.KeyboardEvent){ 
+            continue
+        }
+        switch e.ID {
+            // if user input q or control-c, exit the program. 
+            case "q","<C-c>": 
+    			ui.Close()
+                return -1, nil
+            // if user hit enter means he did his choise, so check waht the input is. if the input is not acceptable, show a warning.
+            case "<Enter>": 
+                // this tempFlag is for recognizing if I need to clear this paragraph, so that user can input a new choice. And it is also for clearing warnings
+                if(tempFlag == false){ 
+                    p1.Text = ""
+                    tempFlag = true
+                }
+                // user input nothing means default choice is chosen
+                if(p1.Text=="") { 
+                    for i, en := range entries {
+                        if en.IsDefault() {
+    						ui.Close()
+                            return i, nil
+		                }
+	                }
+
+    				ui.Close()
+	                //if no one is default, choose the first one. 
+                    return 0, nil 
+                }
+                // try convert the input into a vilid number
+                c, err := strconv.Atoi(p1.Text) 
+                if(err!=nil || c<1 || c>len(entries)){
+                    p1.Text = fmt.Sprintf("%s is not a valid entry number.", p1.Text)
```

Comment:
I might not be understanding how termui works, but why do we need to use the same paragraph for user input and for displaying errors? Would it make sense to have a separate paragraph widget for that, and then on Enter, if the user input is invalid, call a separate function to display the error text, and reset the user input paragraph to ""?

Reply:
I think that way you could use your separate GetInput function, maybe makes this cleaner